pkg/ring/kv/memberlist/memberlist_client.go

```diff
@@ -0,0 +1,851 @@
+	f.DurationVar(&cfg.GossipInterval, prefix+"mlclient.gossip-interval", 0, "How often to gossip. Uses memberlist LAN defaults if 0")
+	f.IntVar(&cfg.GossipNodes, prefix+"mlclient.gossip-nodes", 0, "How many nodes to gossip to. Uses memberlist LAN defaults if 0")
+	f.DurationVar(&cfg.PushPullInterval, prefix+"mlclient.pullpush-interval", 0, "How often to use pull/push mechanism. Uses memberlist LAN defaults if 0")
+	f.DurationVar(&cfg.TimestampInterval, prefix+"mlclient.timestamp-interval", 5*time.Second, "If not 0, update and gossip internal timestamp using this interval. Used by metrics.")
+	f.BoolVar(&cfg.TransportDebug, prefix+"mlclient.transport-debug", false, "Log debug transport messages. Note: global log.level must be at debug level as well.")
+}
+
+// Client implements kv.Client interface by using memberlist gossiping library.
+type Client struct {
+	logger       log.Logger
+	cfg          Config
+	memberlist   *memberlist.Memberlist
+	broadcasts   *memberlist.TransmitLimitedQueue
+	defaultCodec codec.Codec
+
+	// KV Store.
+	storeMu sync.Mutex
+	store   map[string]valueDesc
+
+	// Key watchers
+	watchersMu     sync.Mutex
+	watchers       map[string][]chan string
+	prefixWatchers map[string][]chan string
+
+	// closed on shutdown
+	shutdown chan struct{}
+
+	// metrics
+	numberOfReceivedMessages            prometheus.Counter
+	totalSizeOfReceivedMessages         prometheus.Counter
+	numberOfInvalidReceivedMessages     prometheus.Counter
+	numberOfPulls, numberOfPushes       prometheus.Counter
+	totalSizeOfPulls                    prometheus.Counter
+	totalSizeOfPushes                   prometheus.Counter
+	numberOfBroadcastMessagesInQueue    prometheus.GaugeFunc
+	totalSizeOfBroadcastMessagesInQueue prometheus.Gauge
+	casAttempts                         prometheus.Counter
+	casFailures                         prometheus.Counter
+	casSuccesses                        prometheus.Counter
+	watchPrefixDroppedNotifications     *prometheus.CounterVec
+
+	storeValuesDesc *prometheus.Desc
+	storeSizesDesc  *prometheus.Desc
+
+	memberlistMembersCount    prometheus.GaugeFunc
+	memberlistHealthScore     prometheus.GaugeFunc
+	memberlistMembersInfoDesc *prometheus.Desc
+}
+
+type valueDesc struct {
+	// We store bytes here. Reason is that clients calling CAS function will modify the object in place,
+	// but unless CAS succeeds, we don't want those modifications to be visible.
+	value []byte
+
+	// version is used to detect changes when doing CAS
+	version uint
+
+	// Coded used to encode/decode the value
+	codec codec.Codec
+}
+
+// NewMemberlistClient creates new Client instance. If cfg.JoinMembers is set, it will also try to connect
+// to these members and join the cluster. If that fails and AbortIfJoinFails is true, error is returned and no
+// client is created.
+func NewMemberlistClient(cfg Config, codec codec.Codec) (*Client, error) {
+	logger := cfg.Logger
```

Comment:
I've developed memberlist client outside of Cortex first, this is a leftover from that.

Reply:
Personally, I think it's a good practice to inject dependencies this way, instead of reaching to global logger value, but I'll change the code to fit existing Cortex codebase, and not introduce new patterns. Sorry for that.

Reply:
Haha, I agree with you, no need to be sorry. But we are stuck with this pattern and it's best to be consistent everywhere.

Reply:
Done.